flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/scala/DataStreamConversions.scala

```diff
@@ -59,5 +59,38 @@ class DataStreamConversions[T](dataStream: DataStream[T], inputType: TypeInforma
     }
   }
 
+  /**
+    * Converts the [[DataStream]] with upsert messages into a [[Table]] with keys.
+    *
+    * The message will be encoded as [[Tuple2]]. The first field is a [[Boolean]] flag, the second
+    * field holds the record. A true [[Boolean]] flag indicates an update message, a false flag
+    * indicates a delete message.
+    *
+    * The field name and key of the new [[Table]] can be specified like this:
+    *
+    * {{{
+    *   val env = StreamExecutionEnvironment.getExecutionEnvironment
+    *   val tEnv = TableEnvironment.getTableEnvironment(env)
+    *
+    *   val stream: DataStream[(Boolean, (String, Int))] = ...
+    *   val table = stream.toKeyedTable(tEnv, 'name.key, 'amount)
+    * }}}
+    *
+    * If field names are not explicitly specified, names are automatically extracted from the type
+    * of the [[DataStream]].
+    * If keys are not explicitly specified, an empty key will be used and the table will be a
+    * single row table.
+    *
+    * @param tableEnv The [[StreamTableEnvironment]] in which the new [[Table]] is created.
+    * @param fields The field names of the new [[Table]] (optional).
+    * @return The resulting [[Table]].
+    */
+  def toKeyedTable(tableEnv: StreamTableEnvironment, fields: Expression*): Table = {
```

Comment:
This is some naming inconsistency: `toKeyedTable` vs `fromUpsertStream`. Rename `keyed` to `upsert`?

Reply:
I think we can't use upsert to describe a Table. How about rename it to `toTableFromUpsert`?

Reply:
Maybe `upsertStreamToTable`?

Reply:
I prefer the name starts with toTable. So the methods would be: `toTable`, `toTableFromUpsertStream`, `toTableFromRetractStream`. We can find these methods easier, similar to `fromUpsertStream `, `fromAppendStream `, `fromDataSet `. What do you think?

Reply:
I think the difference between `toTableFromRetractStream` and `toTable` is `toTable` means `toTableFromAppendStream`. The method names are consistence.
Methods in `StreamTableEnvironment`:
- fromAppendStream
- fromUpsertStream
- fromRetractStream

Methods in `DataStreamConversions`:
- toTableFromAppendStream
- toTableFromUpsertStream
- toTableFromRetractStream

Renaming `fromDataStream` to `fromAppendStream` is consistent with renaming `toDataStream` to `toAppendStream` by [FLINK-6543](https://issues.apache.org/jira/browse/FLINK-6543)